### PR TITLE
Implementing the Decorator design pattern

### PR DESCRIPTION
**Pull request by Yemisirach**
- Check the usage of [Decorator pattern in Ruby](https://refactoring.guru/design-patterns/decorator/ruby/example#example-0)
- use two decorators in order to capitalize and trim people's names.
- In the following steps you will get instructions for implementing the Decorator design pattern for this case.
#### Interface
- Create a class `Nameable`.
   - Implement a method called `correct_name` that will raise a `NotImplementedError`.
#### Turn your Person class to Nameable
- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method `correct_name` implemented. It should simply return the name attribute.
#### Prepare base Decorator
- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the `correct_name` method that returns the result of the correct_name method of the `@nameable`.
#### Prepare CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name
+    raise NotImplementedError
+  end
```
Comment (mikethreels) [inline nameable.rb]: - [ ] When raising an error the user might need to know in what class/method the error was, therefore you might want to add a custom error message as you can see below:
```
    def correct_name
        raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

```

File: base_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+require_relative 'nameable'
+
+class BaseDecorator < Nameable
+  def initialize(nameable)
+    super()
+    @nameable = nameable
+  end
```
Comment (mikethreels) [inline base_decorator.rb]: - [ ] please make sure to make the `@nameable` instance variable available outside this class by adding the attribute accessor.

File: trimmer_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+require_relative 'person'
+
+require_relative 'base_decorator'
+
+require_relative 'capitalize_decorator'
+
+class TrimmerDecorator < BaseDecorator
+  def correct_name
+    super().slice(0, 10)
+  end
```
Comment (mikethreels) [inline trimmer_decorator.rb]: - [ ] When you use `super()` in this line you will get the below error:
afbeelding
to fix this please use the instance variable `@nameable` and call the correct_name method like below, and also make sure to first check if the name is longer than 10 or not.
```
 def correct_name
      @nameable.correct_name.length > 10 ? @nameable.correct_name[0..9] : @nameable.correct_name
    end

```

File: capitalize_decorator.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+require_relative 'base_decorator'
+
+require_relative 'person'
+
+class CapitalizeDecorator < BaseDecorator
+  def correct_name
+    super().capitalize!
+  end
```
Comment (mikethreels) [inline capitalize_decorator.rb]: - [ ] When you use `super()` in this line you will get the below error:
afbeelding
to fix this please use the instance variable `@nameable` and call the correct_name method like below, and also make sure to first check if the name is longer than 10 or not.
```
 def correct_name
      @nameable.correct_name.length > 10 ? @nameable.correct_name[0..9] : @nameable.correct_name
    end

```

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Yemi, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- linters are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by gemmen29 — APPROVED** [review]
Hi @Yemisirach, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**